akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -3726,4 +3726,64 @@ trait FlowOpsMat[+Out, +Mat] extends FlowOps[Out, Mat] {
   def monitor: ReprMat[Out, (Mat, FlowMonitor[Out])] =
     monitorMat(Keep.both)
 
+  /**
+   * This is a more scalable and general case of [[groupedWeightedWithin]]
+   * which groups a stream into vectors based on custom weight and time.
+   * The problem with that solution is each grouped vector must fit into memory before emitting to the next stage.
+   * That won't work for the use case of writing large files.
+   * The desirable behavior is to write data as they come as opposed to accumulate everything until the group slicing condition is met.
+   * In this case, the output channel needs to be closed if there is no data arriving within certain time to avoid connection timeout.
+   * This custom flow uses custom aggregator to support such use cases.
+   * Upstream inputs are continuously aggregated as they arrive.
+   * The currently aggregated elements can be terminated and emitted based on custom conditions
+   *
+   * @param allocate    allocate the initial data structure for aggregated elements
+   * @param aggregate   update the aggregated elements, return true if ready to emit after update
+   * @param emitOnTimer decide whether the currently aggregated elements can be emitted on each timer event
+   * @param harvest     this is invoked before emit within the current stage/operator
+   */
+  def aggregateWithBoundary[Agg, Emit](allocate: => Agg)(
+      aggregate: (Agg, Out) => Boolean,
+      harvest: Agg => Emit,
+      emitOnTimer: Option[(Agg => Boolean, FiniteDuration)]): Repr[Emit] =
+    via(AggregateWithBoundary(() => allocate, aggregate, harvest, emitOnTimer))
+
+  /**
+   * This is a convenient wrapper of [[aggregateWithBoundary]] to handle 2 additional time constraints
```

Comment:
Still not 100% sure I agree we should have this quite use case specific overload in the API when we have the more generic one making it possible to build things like this for users. 

Perhaps it could be shown as a usage sample in the reference doc page for the operator instead?

Reply:
this is for the original use case, and it's kind of non-trivial to implement this correctly and efficiently on top of the generic abstraction.

Reply:
I agree that it is non trivial, but I'm not convinced it is generic enough that it _must_ be a part of the API as long as it is now possible (and easier) than before. 

Anything non-trivial added to Akka is also something non-trivial that will need to be maintained for the foreseeable future, so it must bring value for many users.

I'm convinced the general abstraction has such value but the specific two-timeout one, less so.

Reply:
I see your points. I can move the time boundary specific methods to test only for this PR. But I think there needs to be more follow ups to consolidate or restructure some other operators.
The general abstraction created here is probably the "lowest level", very general and minimum composition unit.
The one with 2 timeout in at a "middle level", being a special case for a class of common use cases. Even this middle layer one is a more general abstraction of some existing methods (Grouped+Weighted+Within).
The `FlowOps` and `FlowOpsMat` traits, by my understanding, is supposed to expose usable public methods at all levels of abstraction.
Just from pure theoretic point of view, it's hard to believe skipping a middle layer in the collection of public methods is a consistent decision.

Reply:
I agree in the long run it would be good to express some of the other operators with this intermediate-low level abstraction where possible, but those pre-existing operators has been battle hardened for years, so we'd want this new implementation to get quite a bit of use to find and fix any potential issues not found during review before that is something worth doing.

My argument against the two-timeouts from your use case as a pre-packaged composition of this new operator is not about that in general a higher abstraction level is bad, but that I cannot see that this _specific_ composition is a general enough use case that it should be available from the core stream APIs.